packages/realm-server/migrations/1728561353454_create-user-billing.js

```diff
@@ -0,0 +1,193 @@
+exports.up = (pgm) => {
+  pgm.createTable('users', {
+    id: {
+      type: 'uuid',
+      primaryKey: true,
+      default: pgm.func('gen_random_uuid()'),
+    },
+    matrix_user_id: { type: 'varchar', notNull: true, unique: true },
+    stripe_customer_id: { type: 'varchar' },
+    created_at: {
+      type: 'timestamp',
+      notNull: true,
+      default: pgm.func('current_timestamp'),
+    },
+  });
+
+  pgm.createTable('plans', {
+    id: {
+      type: 'uuid',
+      primaryKey: true,
+      default: pgm.func('gen_random_uuid()'),
+    },
+    name: { type: 'varchar', notNull: true },
+    monthly_price: { type: 'numeric', notNull: true },
+    credits_included: { type: 'integer', notNull: true },
+    created_at: {
+      type: 'timestamp',
+      notNull: true,
+      default: pgm.func('current_timestamp'),
+    },
+  });
+
+  pgm.createTable('subscription_cycles', {
+    id: {
+      type: 'uuid',
+      primaryKey: true,
+      default: pgm.func('gen_random_uuid()'),
+    },
+    subscription_id: { type: 'uuid' },
+    period_start: { type: 'timestamp', notNull: true },
+    period_end: { type: 'timestamp', notNull: true },
+  });
+
+  pgm.createType('subscription_status', ['active', 'canceled', 'expired']);
+
+  pgm.createTable('subscriptions', {
+    id: {
+      type: 'uuid',
+      primaryKey: true,
+      default: pgm.func('gen_random_uuid()'),
+    },
+    user_id: { type: 'uuid', references: 'users(id)' },
+    plan_id: { type: 'uuid', references: 'plans(id)' },
+    started_at: { type: 'timestamp', notNull: true },
+    ended_at: { type: 'timestamp', notNull: true },
+    status: { type: 'subscription_status', notNull: true },
+    stripe_subscription_id: { type: 'varchar', notNull: true },
+    current_billing_cycle_id: {
+      type: 'uuid',
+      references: 'subscription_cycles(id)',
+    },
+  });
+
+  pgm.addConstraint(
+    'subscription_cycles',
+    'subscription_cycles_subscription_id_fkey',
+    {
+      foreignKeys: {
+        columns: 'subscription_id',
+        references: 'subscriptions(id)',
+      },
+    },
+  );
+
+  pgm.createTable('ai_actions', {
+    id: {
+      type: 'uuid',
+      primaryKey: true,
+      default: pgm.func('gen_random_uuid()'),
+    },
+    user_id: { type: 'uuid', references: 'users(id)' },
+    cost_in_usd: { type: 'numeric', notNull: true },
+    created_at: {
+      type: 'timestamp',
+      notNull: true,
+      default: pgm.func('current_timestamp'),
+    },
+  });
+
+  pgm.createType('balance_change_type', [
+    'plan_allowance',
+    'extra_credit_balance',
+  ]);
+
+  pgm.createTable('credit_balance_changes', {
```

Comment:
We will query this table in the following way:

1. Used credits for current subscription period: sum of `change_in_credits` where `balance_change_type` = `plan_allowance` and `billing_cycle_id` = current billing cycle id. We can subtract this from the plan's allowance and see how many credits are left for this period

2. Available credits from the user's balance (this is when users is topping up its credits by paying manually using stripe payment links):  sum of `change_in_credits` where `balance_change_type` = `extra_credit_balance` 

Putting this together will tell us whether user has any credit left to perform more AI actions.

Reply:
We will need a mechanism at the end of the bot streaming response which:

1. Inserts a record in `ai_actions` with the dollar cost of the AI action
2. Converts the cost of ai action to credits 
3. Inserts a record in `credit_balance_changes` (with the foreign id key of `ai_actions`)

Reply:
maybe a better name is `credit_balance_events`.

Reply:
I'm ok with that. I can change it